### ISSUE TITLE
SHA extensions

### ISSUE THREAD

**Issue by jasondavies** [rust-lang-nursery/stdsimd#392]
What needs to be done to support Intel's SHA extensions?

Happy to help if someone is willing to mentor/guide me!

I would also note that there are equivalent/similar extensions for ARM and POWER8.  But getting support for the Intel extensions would be a great start; they are already supported on AMD Ryzen.

**Comment by gnzlbg** [comment]
I can mentor you. It is easier if you report after each step is done, but I am going to give you all the steps necessary here so that you can go at your own pace.

* Step 1: white-list the feature in rustc:
    * find the name of the cpuid bit feature / features, and how to query them
    * find the name of the LLVM backend feature / feature names - `llc --mcpu=help` gives you a least of all supported cpus by your LLVM installation, choose one that has the feature and write `llc --mcpu=<cpu_here> --mattr=help` to list all the features that the CPU supports. Find the ones you care about
    * come up with good names to name these features in rustc - the cpuid bit name, the llvm name, and or how clang and gcc name them are good candidates
    * white-list the feature in rustc by adding it to this list: https://github.com/rust-lang/rust/blob/36b66873187e37a9d79adad89563088a9cb86028/src/librustc_trans/llvm_util.rs#L88
   * submit a PR to rustc - you can work on the next steps but probably this PR will need to land first before any of your code will compile - this can take from a couple of hours to a couple of days, depending on how many PRs are on flight
* Step 2: implementing run-time feature detection for the intrinsic in `stdsimd`
    * you need to add the intrinsic to the supported x86 intrinsics in the `detect` module, for `x86` that would be here: https://github.com/rust-lang-nursery/stdsimd/blob/master/stdsimd/arch/detect/arch/x86.rs#L28
    * you need to check whether the intrinsic is enabled at run-time, and if that's the case, enable the feature. For x86's you need to check whether the cpuid bit of the intrinsic is set and if so, enable the intrinsic like the other code here does: https://github.com/rust-lang-nursery/stdsimd/blob/master/stdsimd/arch/detect/os/x86.rs#L125
* Step 3: implementing the intrinsic
  * come up with a place to implement the feature inside the `coresimd/x86` module. If the feature is its own self-standing feature, you probably want to add a new module there called `<featurename>.rs` 
  * you now have to add the intrinsics to this module, check how that's done for other modules - typically you have to include an `extern` LLVM intrinsic mapping it to rust types (like here: https://github.com/rust-lang-nursery/stdsimd/blob/master/coresimd/x86/rdtsc.rs#L48) and then write a rust wrapper over it (like here: https://github.com/rust-lang-nursery/stdsimd/blob/master/coresimd/x86/rdtsc.rs#L22)
  * finding the names of the llvm intrinsics is tricky, you can search for them in https://github.com/llvm-mirror/clang/tree/master/lib/Headers to see how clang implements them - if it uses a builtin, you can search for some part of the builtin name in clang and/or the llvm sources: https://github.com/llvm-mirror/llvm, and you might land on some tests or some code that tells you how the intrinsic is called, like here: https://github.com/llvm-mirror/clang/blob/a38ba9770a70056f9fdd6c71f819e5db45a105e4/test/CodeGen/rd-builtins.c#L15
  * you need to add a `assert_instr(instruction name)` that will check for you in CI that the intrinsic does actually generate the assembly instruction that it is supposed to
  * you need to add a test module (check the rdtsc module) that contains `#[simd_test = "feature-name"]` that exercise each intrinsic at run-time and check that they do produce the correct results
  * Intrinsics need docs: one sentence stating what the intrinsic does is fine. You can mention some corner cases if you want - there are many sources of how other compilers (clang, msvc) document intrinsics to get inspiration from. Keep in mind that this library provides the intrinsics, but at least at this point in time is not a tutorial on how to use them correctly. Currently, we only have examples for intrinsics that can do way to many things (like the string manipulation ones) or have way too many corner cases, but these are very very rare.   
* Step 4: get test to pass on CI
  * test might fail on x86/x86-64 targets for a variety of reasons... we have already seen many of them and will be able to help you a lot here, so go directly to step 5.  
* Step 5: submit a PR to stdsimd!
  * If you get stuck or need help submit a PR early and we can give you feedback as you progress.

**Comment by alexcrichton** [comment]
Oh if you'd like I got started on these [awhile ago](https://gist.github.com/alexcrichton/943a64e12e3affd7a40aa7eeb0775bd2) but never got around to adding tests and landing them, so feel free to use that as a starting point!

**Comment by jasondavies** [comment]
Wonderful.  Thank you!

* Step 1: unless I'm missing something, it seems the `sha` feature is already white-listed in rustc?  Should save a bit of time if so!

**Comment by gnzlbg** [comment]
> Step 1: unless I'm missing something, it seems the sha feature is already white-listed in rustc? Should save a bit of time if so!

Yes, `sha` is already white-listed. If that's everything you need then you can skip step 1 and go straight to step 2 and take it where @alexcrichton left it. 

Also, the `AArch64` `crypto` feature is white-listed and it stands for `sha1`, `sha2`, `aes`, and `pmull`: https://github.com/rust-lang-nursery/stdsimd/blob/master/stdsimd/arch/detect/arch/aarch64.rs#L91 , so if that's all you need you can go directly to step 3 for `aarch64` as well. AArch64's also has `sha3` and `sha512` features but those aren't white-listed yet (also I don't know whether LLVM supports these yet; it should though).

**Comment by jasondavies** [comment]
I'll look at adding `aarch64` if I can get hold of a test machine.  Will probably be a separate PR.

**Comment by gnzlbg** [comment]
An alternative would be to get hold of qemu-aarch64 or use our CI via travis to implement these.

**Comment by jasondavies** [comment]
I got hold of an ARMv8 server (with 96 cores!) but I'm running into some strange errors:

```
error:
        is_x86_feature_detected can only be used on x86 and x86_64 targets.
        You can prevent it from being used in other architectures by
        guarding it behind a cfg(target_arch) as follows:

            #[cfg(any(target_arch = "x86", target_arch = "x86_64"))] {
                if is_x86_feature_detected(...) { ... }
            }

   --> crates/coresimd/src/../../../coresimd/aarch64/neon.rs:725:5
    |
725 |     #[simd_test = "neon"]
    |     ^^^^^^^^^^^^^^^^^^^^^
    |
    = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```

Any idea what's going on here?  Tested on current master (56937d25fb6e4f3fc14c9f0e29123050cac96aca) with `rustc 1.26.0-nightly (75af15ee6 2018-03-20)`.

**Comment by jasondavies** [comment]
Forgot to say: there are 67 of these errors generated when running `cargo test`.  This is on Linux (Ubuntu 16.04 LTS).

**Comment by jasondavies** [comment]
The system is a dual-socket Cavium ThunderX SoC.

**Comment by alexcrichton** [comment]
@jasondavies if you set a `TARGET` environment variable to the target you're compiling for, does that fix the issue?

**Comment by jasondavies** [comment]
Yes, it does!

**Comment by alexcrichton** [comment]
Ah ok, I think the bug is here --

https://github.com/rust-lang-nursery/stdsimd/blob/56937d25fb6e4f3fc14c9f0e29123050cac96aca/crates/simd-test-macro/src/lib.rs#L73-L75

I think Cargo doesn't actually set that for crate compilations, so we should fix that!

**Comment by alexcrichton** [comment]
Ok I think I've fixed that in https://github.com/rust-lang-nursery/stdsimd/pull/397 and otherwise this should be fixed, so closing.

**Comment by gnzlbg** [comment]
Ah, I see. So my opinion on this and #397 is that somebody checking out `stdsimd` and running `ci/run.sh` or maybe even just `cargo test` should get a "nice" out-of-the-box experience. At least with respect to being able to run the test on the platform they are developing on. I think this is important to get good bug reports.

So the code that #397 fixes actually tried to come up with sensible defaults for linux, macosx, and windows, but it might well be that reasonable defaults do not really exist. If that's the case erroring early sounds like a good plan to me.
